dask/array/tests/test_slicing.py

```diff
@@ -530,7 +530,63 @@ def test_oob_check():
         x[0, 0]
 
 
-def test_index_with_dask_array():
+@pytest.mark.parametrize('idx_chunks', [None, 3, 2, 1])
+@pytest.mark.parametrize('x_chunks', [None, (3, 5), (2, 3), (1, 2), (1, 1)])
+def test_index_with_int_dask_array(x_chunks, idx_chunks):
+    # test data is crafted to stress use cases:
+    # - pick from different chunks of x out of order
+    # - a chunk of x contains no matches
+    # - only one chunk of x
+    x = np.array([[10, 20, 30, 40, 50],
+                  [60, 70, 80, 90, 100],
+                  [110, 120, 130, 140, 150]])
+    idx = np.array([3, 0, 1])
+    expect = np.array([[40, 10, 20],
+                       [90, 60, 70],
+                       [140, 110, 120]])
+
+    if x_chunks is not None:
+        x = da.from_array(x, chunks=x_chunks)
+    if idx_chunks is not None:
+        idx = da.from_array(idx, chunks=idx_chunks)
+
+    assert_eq(x[:, idx], expect)
+    assert_eq(x.T[idx, :], expect.T)
+
+
+@pytest.mark.parametrize('chunks', [1, 2, 3])
+def test_index_with_int_dask_array_0d(chunks):
+    # Slice by 0-dimensional array
+    x = da.from_array(np.array([[10, 20, 30],
+                                [40, 50, 60]]), chunks=chunks)
+    idx0 = da.from_array(np.array(1), chunks=1)
+    assert_eq(x[idx0, :], x[1, :])
+    assert_eq(x[:, idx0], x[:, 1])
+
+
+@pytest.mark.parametrize('chunks', [1, 2, 3, 4, 5])
+def test_index_with_int_dask_array_nanchunks(chunks):
+    # Slice by array with nan-sized chunks
+    a = da.arange(-2, 3, chunks=chunks)
+    assert_eq(a[a.nonzero()], np.array([-2, -1,  1,  2]))
+    # Edge case: the nan-sized chunks resolve to size 0
+    a = da.zeros(5, chunks=chunks)
+    assert_eq(a[a.nonzero()], np.array([]))
+
+
+@pytest.mark.parametrize('chunks', [2, 4])
+def test_index_with_int_dask_array_indexerror(chunks):
+    # Slice by array with nan-sized chunks
+    a = da.arange(4, chunks=chunks)
+    idx = da.from_array(np.array([-1]), chunks=1)
+    with pytest.raises(IndexError):
+        a[idx].compute()
```

Comment:
I don't understand this test -- shouldn't indexing by -1 for the last element be valid?

Can we normalize integers indices first (by adding the size of the dimension if necessary) to ensure they're in the valid range?

Reply:
fixed - added support for negative indices, with explicit unit tests